core/src/hir/lowering.rs

```diff
@@ -842,6 +842,16 @@ impl<'ast, 'errors> LoweringContext<'ast, 'errors> {
                     _ => None,
                 }
             }
+            ty @ ast::TypeName::Option(value_ty) => {
+                if let Some(ty) = self.lower_out_type(ty, return_ltl.as_mut(), in_path) {
+                    // reference type
+                    Some(ReturnType::Infallible(Some(SuccessType::OutType(ty))))
+                } else {
+                    // value type
+                    self.errors.remove(self.errors.len() - 1);
```

Comment:
issue: no let's not do that

Reply:
I think this will need a bit of a refactor of `lower_out_type` so that it:

 - takes in an `optional` bool parameter
 - doesn't itself check for `Option` anymore (the caller sets `optional`)

and then here we should just check if the inner type is a reference/box.

Reply:
It does have to check for `Option` itself if we want to allow things like `Result<Option<Box<Foo>>, ()>`

Reply:
We can retain that check, we should not be removing errors

Reply:
It's removing the single error that the failed invocation in line 846 added. Because that's not actually an error, it should not show up in the list of errors if there's a later, fatal error.

Reply:
Ah, I think I see where we differ. I was hoping for non-reference Option support to be more generic than just in the outset of return values but that's not what happens here so this somewhat ugly dance of responsibilities still needs to be done.

In that case can we still have that check external here: have it check for `Reference|Box` against `value_ty`?